### ISSUE TITLE
Worse performance when using megatron-bridge for LoRA training compared to FSDP2

### ISSUE THREAD

**Issue by acphile** [verl-project/verl#5094]
### System Info

----------Python Info----------
Version      : 3.12.12
Compiler     : GCC 14.3.0
Build        : ('main', 'Oct 22 2025 23:25:55')
Arch         : ('64bit', 'ELF')
------------Pip Info-----------
Version      : 25.3
vllm         : 0.11.0
sglang       : not found.
ray          : 2.53.0
torch        : 2.8.0
----------verl Info-----------
Version      : 0.7.0.dev
Commit Hash  : f9c855f7cf04d603c9546bc01776c74806a879c1
----------Platform Info----------
Platform     : Linux-6.8.0-1029-aws-x86_64-with-glibc2.35
system       : Linux
version      : #31~22.04.1-Ubuntu SMP Thu Apr 24 21:16:18 UTC 2025
----------Environment----------
OMP_NUM_THREADS="1"
CUDA Runtime : 12.8
CUDA Compiler : Cuda compilation tools, release 12.6, V12.6.85

Megatron-bridge: use the commit 83a7c1134c562d8c6decd10a1f0a6e6a7a8a3a44 as suggested in https://verl.readthedocs.io/en/latest/advance/ppo_lora.html
Megatron-core: megatron_core-0.15.2
transformer_engine: 2.6.0+torch2.8.0cu128

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction
I use the standard GSM8K dataset.

The LoRA training script based on `megatron-bridge`
```
project_name=sanity_check
TP=2
PP=1
EP=1
ETP=1
CP=1
lr=1e-5
experiment_name=qwen3_4b_lora_megatron_${PP}_${TP}_${EP}_noeager_lr_${lr}
mkdir -p outputs/$project_name/$experiment_name/

export VERL_LOGGING_LEVEL=INFO
export VLLM_LOGGING_LEVEL=INFO
export PYTHONUNBUFFERED=1
max_token_len_per_gpu=32768 # 16384

python3 -m verl.trainer.main_ppo \
    --config-path=config \
    --config-name='ppo_megatron_trainer.yaml' \
    algorithm.adv_estimator=grpo \
    data.train_files=$HOME/data/gsm8k/train.parquet \
    data.val_files=$HOME/data/gsm8k/test.parquet \
    data.train_batch_size=128 \
    data.max_prompt_length=512 \
    data.max_response_length=1024 \
    data.filter_overlong_prompts=True \
    data.truncation='error' \
    data.shuffle=True \
    actor_rollout_ref.model.path=Qwen/Qwen3-4B \
    actor_rollout_ref.model.lora.rank=32 \
    actor_rollout_ref.model.lora.alpha=64 \
    actor_rollout_ref.model.use_fused_kernels=True \
    actor_rollout_ref.model.lora.lora_A_init_method="kaiming" \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.actor.optim.lr=${lr} \
    actor_rollout_ref.actor.ppo_mini_batch_size=32 \
    actor_rollout_ref.actor.use_dynamic_bsz=True \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=2 \
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=4 \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=4 \
    actor_rollout_ref.actor.ppo_max_token_len_per_gpu=${max_token_len_per_gpu} \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.kl_loss_coef=0.001 \
    actor_rollout_ref.actor.kl_loss_type=low_var_kl \
    actor_rollout_ref.actor.entropy_coeff=0 \
    actor_rollout_ref.actor.megatron.use_mbridge=True \
    actor_rollout_ref.actor.megatron.vanilla_mbridge=False \
    actor_rollout_ref.actor.megatron.tensor_model_parallel_size=${TP} \
    actor_rollout_ref.actor.megatron.pipeline_model_parallel_size=${PP} \
    actor_rollout_ref.actor.megatron.expert_model_parallel_size=${EP} \
    actor_rollout_ref.actor.megatron.context_parallel_size=${CP} \
    actor_rollout_ref.actor.megatron.expert_tensor_parallel_size=${ETP} \
    actor_rollout_ref.actor.megatron.sequence_parallel=False \
    actor_rollout_ref.actor.megatron.param_offload=False \
    actor_rollout_ref.actor.megatron.optimizer_offload=False \
    actor_rollout_ref.actor.megatron.grad_offload=False \
    +actor_rollout_ref.actor.megatron.override_transformer_config.recompute_method=uniform \
    +actor_rollout_ref.actor.megatron.override_transformer_config.recompute_granularity=full \
    +actor_rollout_ref.actor.megatron.override_transformer_config.recompute_num_layers=1 \
    actor_rollout_ref.ref.megatron.tensor_model_parallel_size=${TP} \
    actor_rollout_ref.ref.megatron.pipeline_model_parallel_size=${PP} \
    actor_rollout_ref.ref.megatron.expert_model_parallel_size=${EP} \
    actor_rollout_ref.ref.megatron.context_parallel_size=${CP} \
    actor_rollout_ref.ref.megatron.expert_tensor_parallel_size=${ETP} \
    actor_rollout_ref.ref.megatron.sequence_parallel=False \
    actor_rollout_ref.ref.megatron.param_offload=True \
    actor_rollout_ref.rollout.log_prob_use_dynamic_bsz=True \
    actor_rollout_ref.rollout.log_prob_max_token_len_per_gpu=${max_token_len_per_gpu} \
    actor_rollout_ref.rollout.tensor_model_parallel_size=2 \
    actor_rollout_ref.rollout.name=vllm \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.5 \
    actor_rollout_ref.rollout.n=8 \
    actor_rollout_ref.rollout.enforce_eager=False \
    actor_rollout_ref.rollout.free_cache_engine=True \
    actor_rollout_ref.ref.log_prob_use_dynamic_bsz=True \
    actor_rollout_ref.ref.log_prob_max_token_len_per_gpu=${max_token_len_per_gpu} \
    algorithm.use_kl_in_reward=False \
    trainer.critic_warmup=0 \
    trainer.logger=['console','tensorboard'] \
    trainer.project_name=$project_name \
    trainer.experiment_name=$experiment_name \
    trainer.n_gpus_per_node=8 \
    trainer.nnodes=1 \
    trainer.save_freq=20 \
    trainer.test_freq=10 \
    trainer.total_epochs=1 
```

The LoRA training script based on `FSDP2`:
```
python3 -m verl.trainer.main_ppo \
    algorithm.adv_estimator=grpo \
    data.train_files=$HOME/data/gsm8k/train.parquet \
    data.val_files=$HOME/data/gsm8k/test.parquet \
    data.train_batch_size=128 \
    data.max_prompt_length=512 \
    data.max_response_length=1024 \
    data.filter_overlong_prompts=True \
    data.truncation='error' \
    data.shuffle=True \
    actor_rollout_ref.model.path=Qwen/Qwen3-4B \
    actor_rollout_ref.model.lora_rank=32 \
    actor_rollout_ref.model.lora_alpha=64 \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.actor.optim.lr=1e-5 \
    actor_rollout_ref.actor.ppo_mini_batch_size=32 \
    actor_rollout_ref.actor.use_dynamic_bsz=True \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=2 \
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=4 \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=4 \
    actor_rollout_ref.actor.ppo_max_token_len_per_gpu=${max_token_len_per_gpu} \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.kl_loss_coef=0.001 \
    actor_rollout_ref.actor.kl_loss_type=low_var_kl \
    actor_rollout_ref.actor.entropy_coeff=0 \
    actor_rollout_ref.actor.strategy=fsdp2 \
    actor_rollout_ref.model.enable_gradient_checkpointing=True \
    actor_rollout_ref.actor.fsdp_config.param_offload=False \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=False \
    actor_rollout_ref.rollout.log_prob_use_dynamic_bsz=True \
    actor_rollout_ref.rollout.log_prob_max_token_len_per_gpu=${max_token_len_per_gpu} \
    actor_rollout_ref.rollout.tensor_model_parallel_size=2 \
    actor_rollout_ref.rollout.name=vllm \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.5 \
    actor_rollout_ref.rollout.n=8 \
    actor_rollout_ref.rollout.enforce_eager=False \
    actor_rollout_ref.rollout.free_cache_engine=True \
    actor_rollout_ref.ref.log_prob_use_dynamic_bsz=True \
    actor_rollout_ref.ref.log_prob_max_token_len_per_gpu=${max_token_len_per_gpu} \
    algorithm.use_kl_in_reward=False \
    trainer.critic_warmup=0 \
    trainer.logger=['console','tensorboard'] \
    trainer.project_name=$project_name \
    trainer.experiment_name=$experiment_name \
    trainer.n_gpus_per_node=8 \
    trainer.nnodes=1 \
    trainer.save_freq=20 \
    trainer.test_freq=10 \
    trainer.total_epochs=1
```
The only difference lies in the training backend (FSDP2 vs megatron-bridge)

### Expected behavior

The two script should lead to similar performance (`acc/mean@1`). However, training with FSDP2 is always better than training with megatron-bridge in the end:
FSDP2: `0.92~0.94`
Megatron-bridge: `0.82~0.89` (one time it even collapses (0 accuracy))

**Comment by Kirrito-k423** [comment]
This performance gap between megatron-bridge and FSDP2 for LoRA training is concerning. The ~10% accuracy difference and occasional collapse suggest a fundamental issue in the megatron-bridge LoRA implementation.

Potential causes:

1. **LoRA initialization differences**: Megatron-bridge may initialize LoRA weights differently than FSDP2. Check `lora_A_init_method="kaiming"` - this is correct but may not match FSDP2's default.

2. **Gradient accumulation**: With `use_dynamic_bsz=True`, megatron-bridge's gradient accumulation may not be handling LoRA gradients correctly across micro-batches.

3. **Weight synchronization**: The weight sync between megatron training and vLLM rollout may have subtle bugs for LoRA parameters.

4. **Learning rate sensitivity**: Megatron-bridge may need different LR for LoRA. Try `lr=5e-6` or `lr=2e-5` to see if the gap narrows.

**Debugging steps:**

1. **Compare gradient norms**:
```python
# Log gradient norms for LoRA params
for name, param in model.named_parameters():
    if 'lora' in name and param.grad is not None:
        print(f"{name}: grad_norm={param.grad.norm().item()}")
```

2. **Check LoRA rank effectiveness**:
Compare the effective rank of LoRA updates between FSDP2 and megatron-bridge.

3. **Disable dynamic batching temporarily**:
```yaml
actor_rollout_ref.actor.use_dynamic_bsz=False
```

This will help isolate if the issue is in dynamic batching or in LoRA itself.

4. **Profile with wandb**:
Log per-step metrics to compare training dynamics between the two backends.

This warrants investigation from the megatron-bridge team.

**Comment by yaoyu-33** [comment]
Hi @acphile — thanks for reporting this. We've opened a tracking issue on the Megatron Bridge side to investigate: https://github.com/NVIDIA-NeMo/Megatron-Bridge/issues/3274

We'll look into potential causes including LoRA weight initialization differences, gradient scaling with activation recomputation, and weight sync behavior in the Bridge LoRA path.

In the meantime, a few questions that could help us narrow down the root cause:

1. Does the accuracy gap persist if you disable activation recomputation on the Bridge side (remove the `override_transformer_config.recompute_*` overrides)?
2. Have you tried `sequence_parallel=True` with the Megatron backend? (Just to rule out any silent interaction.)
3. For the run that collapsed to 0 accuracy — did you see any warnings or NaN gradients in the logs?

We'll follow up in the Bridge issue as we make progress.

**Comment by HollowMan6** [comment]
Hi @acphile, would you mind letting us know which Megatron-Bridge version you are currently using? We recommend https://github.com/NVIDIA-NeMo/Megatron-Bridge/commit/6259ae83c735c4412796fc5cfb4c9607b949ae29 or later https://github.com/verl-project/verl/blob/516657fa39595dd557ba34b98ae58d25d0f44e66/examples/grpo_trainer/run_qwen2-7b_math_megatron_lora.sh#L6

What you suggested (one time it even collapses (0 accuracy)) sounds exactly like what this PR on the Megatron-Bridge side has fixed: https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/1817:

Before the fix:
<img width="3694" height="1678" alt="Image" src="https://github.com/user-attachments/assets/95209d09-4fdd-4fb8-95b2-53375b401891" />

After the fix:

<img width="3700" height="1664" alt="Image" src="https://github.com/user-attachments/assets/0dd798bb-c04c-4e2c-802e-099087ed1e85" />

**Comment by sajadn** [comment]
@acphile @Kirrito-k423 can you confirm that the issue is resolved?
